File: packages/backend-node/src/domains/file-io.ts
import type {
  DlaDescriptor,
  FileIoApi,
  FoundDlaDescriptor,
  SpiceHandle,
} from "@rybosome/tspice-backend-contract";
import { invariant } from "@rybosome/tspice-core";

import type { NativeAddon } from "../runtime/addon.js";
import type { SpiceHandleRegistry, SpiceHandleKind } from "../runtime/spice-handles.js";
const I32_MIN = -2147483648;
const I32_MAX = 2147483647;

const DAS_BACKED = ["DAS", "DLA"] as const satisfies readonly SpiceHandleKind[];

function assertDlaDescriptor(value: unknown, context: string): asserts value is DlaDescriptor {
  invariant(typeof value === "object" && value !== null, `${context}: expected an object`);
  const obj = value as Record<string, unknown>;
  for (const key of [
    "bwdptr",
    "fwdptr",
    "ibase",
    "isize",
    "dbase",
    "dsize",
    "cbase",
    "csize",
  ] as const) {
    const v = obj[key];
    invariant(
      typeof v === "number" &&
        Number.isInteger(v) &&
        v >= I32_MIN &&
        v <= I32_MAX,
      `${context}: expected ${key} to be a 32-bit signed integer`,
    );

    const min = key === "bwdptr" || key === "fwdptr" ? -1 : 0;
    invariant(
      v >= min,
      `${context}: expected ${key} to be >= ${min}`,
    );
  }
}

function normalizeFoundDlaDescriptor(value: unknown, context: string): FoundDlaDescriptor {
  invariant(typeof value === "object" && value !== null, `${context}: expected an object`);
  const obj = value as { found?: unknown; descr?: unknown };
  invariant(typeof obj.found === "boolean", `${context}: expected found to be a boolean`);

  if (!obj.found) {
    return { found: false };
  }

  assertDlaDescriptor(obj.descr, `${context}.descr`);
  return { found: true, descr: obj.descr };
}

export function createFileIoApi(native: NativeAddon, handles: SpiceHandleRegistry): FileIoApi {
  function closeDasBacked(handle: SpiceHandle, context: string): void {
    handles.close(
      handle,
      DAS_BACKED,
      (entry) => {
        // In CSPICE, `dascls_c` closes both DAS and DLA handles, and `dlacls_c`
        // is just an alias.
        native.dascls(entry.nativeHandle);
      },
      context,
    );
  }

  const api = {
    exists: (path: string) => {
      const exists = native.exists(path);
      invariant(typeof exists === "boolean", "Expected native backend exists() to return a boolean");
      return exists;
    },

    getfat: (path: string) => {
      const result = native.getfat(path);
      invariant(typeof result === "object" && result !== null, "Expected native backend getfat() to return an object");
      const obj = result as { arch?: unknown; type?: unknown };
      invariant(typeof obj.arch === "string", "Expected getfat().arch to be a string");
      invariant(typeof obj.type === "string", "Expected getfat().type to be a string");
      return { arch: obj.arch, type: obj.type };
    },

    dafopr: (path: string) => handles.register("DAF", native.dafopr(path)),
    dafcls: (handle: SpiceHandle) =>
      handles.close(handle, ["DAF"], (e) => native.dafcls(e.nativeHandle), "dafcls"),
    dafbfs: (handle: SpiceHandle) =>
      native.dafbfs(handles.lookup(handle, ["DAF"], "dafbfs").nativeHandle),
    daffna: (handle: SpiceHandle) => {
      const found = native.daffna(handles.lookup(handle, ["DAF"], "daffna").nativeHandle);
      invariant(typeof found === "boolean", "Expected native backend daffna() to return a boolean");
      return found;
    },

    dasopr: (path: string) => handles.register("DAS", native.dasopr(path)),
    dascls: (handle: SpiceHandle) => closeDasBacked(handle, "dascls"),

    dlaopn: (path: string, ftype: string, ifname: string, ncomch: number) =>
      handles.register("DLA", native.dlaopn(path, ftype, ifname, ncomch)),

    dlabfs: (handle: SpiceHandle) =>
      normalizeFoundDlaDescriptor(
        native.dlabfs(handles.lookup(handle, DAS_BACKED, "dlabfs").nativeHandle),
        "dlabfs()",
      ),

    dlafns: (handle: SpiceHandle, descr: DlaDescriptor) => {
      assertDlaDescriptor(descr, "dlafns(descr)");
      return normalizeFoundDlaDescriptor(
        native.dlafns(handles.lookup(handle, DAS_BACKED, "dlafns").nativeHandle, descr),
        "dlafns()",
      );
    },

    dlacls: (handle: SpiceHandle) => closeDasBacked(handle, "dlacls"),

    dskopn: (path: string, ifname: string, ncomch: number) =>
      // DSKs are DAS-backed; register as DAS so `dascls` can close it.
      handles.register("DAS", native.dskopn(path, ifname, ncomch)),

    dskmi2: (
      nv: number,
      vrtces: readonly number[],
      np: number,
      plates: readonly number[],
      finscl: number,
      corscl: number,
      worksz: number,
      voxpsz: number,
      voxlsz: number,
      makvtl: boolean,
      spxisz: number,
    ) => {
      const result = native.dskmi2(
        nv,
        vrtces,
        np,
        plates,
        finscl,
        corscl,
        worksz,
        voxpsz,
        voxlsz,
        makvtl,
        spxisz,
      );
      invariant(
        typeof result === "object" && result !== null,
        "Expected native backend dskmi2() to return an object",
      );
      const obj = result as { spaixd?: unknown; spaixi?: unknown };
      invariant(Array.isArray(obj.spaixd), "Expected dskmi2().spaixd to be an array");
      invariant(Array.isArray(obj.spaixi), "Expected dskmi2().spaixi to be an array");
      invariant(obj.spaixd.every((v) => typeof v === "number"), "Expected dskmi2().spaixd to be a number[]");
      invariant(obj.spaixi.every((v) => typeof v === "number"), "Expected dskmi2().spaixi to be a number[]");
      return { spaixd: obj.spaixd as number[], spaixi: obj.spaixi as number[] };
    },

    dskw02: (
      handle: SpiceHandle,
      center: number,
      surfid: number,
      dclass: number,
      frame: string,
      corsys: number,
      corpar: readonly number[],
      mncor1: number,
      mxcor1: number,
      mncor2: number,
      mxcor2: number,
      mncor3: number,
      mxcor3: number,
      first: number,
      last: number,
      nv: number,
      vrtces: readonly number[],
      np: number,
      plates: readonly number[],
      spaixd: readonly number[],
      spaixi: readonly number[],
    ) =>
      native.dskw02(
        handles.lookup(handle, ["DAS"], "dskw02").nativeHandle,
        center,
        surfid,
        dclass,
        frame,
        corsys,
        corpar,
        mncor1,
        mxcor1,
        mncor2,
        mxcor2,
        mncor3,
        mxcor3,
        first,
        last,
        nv,
        vrtces,
        np,
        plates,
        spaixd,
        spaixi,
      ),
  } satisfies FileIoApi;

  Object.defineProperty(api, "__debugOpenHandleCount", {
    value: () => handles.size(),
    enumerable: false,
  });

  return api;
}

File: packages/py-parity-checking/sidecars/spiceypy/src/py_parity_sidecar/handlers/file_io.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

import numpy as np
import spiceypy as sp
from spiceypy.utils.exceptions import NotFoundError

from ..models import (
    StepFileIoDafbfs,
    StepFileIoDafcls,
    StepFileIoDaffna,
    StepFileIoDafopr,
    StepFileIoDascls,
    StepFileIoDasopr,
    StepFileIoDlabfs,
    StepFileIoDlacls,
    StepFileIoDlafns,
    StepFileIoDlaopn,
    StepFileIoDskmi2,
    StepFileIoDskopn,
    StepFileIoDskw02,
    StepFileIoExists,
    StepFileIoGetfat,
    StepOutput,
    WorkflowStep,
)
from ..runtime import SidecarRuntimeContext, resolve_path_ref

SPAIX_HEAD_LIMIT = 8


@dataclass
class HandleState:
    handle: int
    kind: str
    closeWith: str
    isOpen: bool


@dataclass
class FileIoState:
    handles: dict[str, HandleState]
    descriptors: dict[str, Any]
    spatialIndexes: dict[str, tuple[np.ndarray, np.ndarray]]


def _resolve_file_io_path(context: SidecarRuntimeContext, path_value: str) -> str:
    return resolve_path_ref(context.paths, path_value)


def _prepare_output_path(path_value: str) -> None:
    path = Path(path_value)
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists() and path.is_file():
        path.unlink()


def _register_handle(
    state: FileIoState,
    handle_id: str,
    handle: int,
    *,
    kind: str,
    close_with: str,
) -> None:
    existing = state.handles.get(handle_id)
    if existing is not None and existing.isOpen:
        raise ValueError(f"File-io handle already exists and is open: {handle_id}")

    state.handles[handle_id] = HandleState(
        handle=int(handle),
        kind=kind,
        closeWith=close_with,
        isOpen=True,
    )


def _require_handle(
    state: FileIoState,
    handle_id: str,
    *,
    expected_kind: str | None = None,
    require_open: bool = False,
) -> HandleState:
    handle_state = state.handles.get(handle_id)
    if handle_state is None:
        raise ValueError(f"File-io handle does not exist: {handle_id}")

    if expected_kind is not None and handle_state.kind != expected_kind:
        raise ValueError(
            f"File-io handle {handle_id} has kind {handle_state.kind}, expected {expected_kind}",
        )

    if require_open and not handle_state.isOpen:
        raise ValueError(f"File-io handle is invalid or closed: {handle_id}")

    return handle_state


def _require_descriptor(state: FileIoState, descr_id: str) -> Any:
    descr = state.descriptors.get(descr_id)
    if descr is None:
        raise ValueError(f"File-io descriptor does not exist: {descr_id}")
    return descr


def _require_spatial_index(state: FileIoState, spaix_id: str) -> tuple[np.ndarray, np.ndarray]:
    index = state.spatialIndexes.get(spaix_id)
    if index is None:
        raise ValueError(f"File-io spatial index does not exist: {spaix_id}")
    return index


def _normalize_dla_result(value: Any) -> tuple[bool, Any | None]:
    if isinstance(value, tuple) and len(value) == 2 and isinstance(value[1], (bool, int)):
        descr, found_raw = value
        found = bool(found_raw)
        if not found:
            return False, None
        return True, descr

    return True, value


def _reshape_vrtces(flat_vrtces: list[float], *, nv: int, label: str) -> np.ndarray:
    if nv <= 0:
        raise ValueError(f"{label}.nv must be > 0")

    expected_len = nv * 3
    if len(flat_vrtces) != expected_len:
        raise ValueError(f"{label}.vrtces length must be {expected_len}")

    return np.asarray(flat_vrtces, dtype=np.float64).reshape((nv, 3))


def _reshape_plates(flat_plates: list[int], *, np_count: int, label: str) -> np.ndarray:
    if np_count <= 0:
        raise ValueError(f"{label}.np must be > 0")

    expected_len = np_count * 3
    if len(flat_plates) != expected_len:
        raise ValueError(f"{label}.plates length must be {expected_len}")

    return np.asarray(flat_plates, dtype=np.int32).reshape((np_count, 3))


def _summarize_spatial_index(spaixd: np.ndarray, spaixi: np.ndarray) -> dict[str, Any]:
    return {
        "spaixdLength": int(spaixd.size),
        "spaixiLength": int(spaixi.size),
        "spaixdHead": spaixd[:SPAIX_HEAD_LIMIT].astype(float).tolist(),
        "spaixiHead": spaixi[:SPAIX_HEAD_LIMIT].astype(int).tolist(),
    }


def run_file_io_step(step: WorkflowStep, state: FileIoState, context: SidecarRuntimeContext) -> StepOutput | None:
    if isinstance(step, StepFileIoExists):
        return StepOutput(op=step.op, value=bool(sp.exists(_resolve_file_io_path(context, step.path))))

    if isinstance(step, StepFileIoGetfat):
        arch, ftype = sp.getfat(_resolve_file_io_path(context, step.path))
        return StepOutput(op=step.op, value={"arch": str(arch), "type": str(ftype)})

    if isinstance(step, StepFileIoDafopr):
        handle = int(sp.dafopr(_resolve_file_io_path(context, step.path)))
        _register_handle(state, step.handleId, handle, kind="DAF", close_with="dafcls")
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDafcls):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAF",
            require_open=True,
        )
        sp.dafcls(handle_state.handle)
        handle_state.isOpen = False
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDafbfs):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAF",
            require_open=True,
        )
        sp.dafbfs(handle_state.handle)
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDaffna):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAF",
            require_open=True,
        )
        sp.dafcs(handle_state.handle)
        found = bool(sp.daffna())
        return StepOutput(op=step.op, value=found)

    if isinstance(step, StepFileIoDasopr):
        handle = int(sp.dasopr(_resolve_file_io_path(context, step.path)))
        _register_handle(state, step.handleId, handle, kind="DAS", close_with="dascls")
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDascls):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAS",
            require_open=True,
        )
        sp.dascls(handle_state.handle)
        handle_state.isOpen = False
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDlaopn):
        _prepare_output_path(_resolve_file_io_path(context, step.path))
        handle = int(sp.dlaopn(_resolve_file_io_path(context, step.path), step.ftype, step.ifname, step.ncomch))
        _register_handle(state, step.handleId, handle, kind="DAS", close_with="dascls")
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDlabfs):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAS",
            require_open=True,
        )
        try:
            found, descr = _normalize_dla_result(sp.dlabfs(handle_state.handle))
        except NotFoundError:
            found = False
            descr = None

        if not found or descr is None:
            state.descriptors.pop(step.descrId, None)
            return StepOutput(op=step.op, value={"found": False})

        state.descriptors[step.descrId] = descr
        return StepOutput(op=step.op, value={"found": True})

    if isinstance(step, StepFileIoDlafns):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAS",
            require_open=True,
        )
        descr = _require_descriptor(state, step.descrId)
        try:
            found, next_descr = _normalize_dla_result(sp.dlafns(handle_state.handle, descr))
        except NotFoundError:
            found = False
            next_descr = None

        if not found or next_descr is None:
            state.descriptors.pop(step.descrId, None)
            return StepOutput(op=step.op, value={"found": False})

        state.descriptors[step.descrId] = next_descr
        return StepOutput(op=step.op, value={"found": True})

    if isinstance(step, StepFileIoDlacls):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAS",
            require_open=True,
        )
        # SpiceyPy does not expose dlacls; CSPICE defines it as an alias of dascls.
        sp.dascls(handle_state.handle)
        handle_state.isOpen = False
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDskopn):
        _prepare_output_path(_resolve_file_io_path(context, step.path))
        handle = int(sp.dskopn(_resolve_file_io_path(context, step.path), step.ifname, step.ncomch))
        _register_handle(state, step.handleId, handle, kind="DAS", close_with="dascls")
        return StepOutput(op=step.op, value=None)

    if isinstance(step, StepFileIoDskmi2):
        vrtces = _reshape_vrtces(step.vrtces, nv=step.nv, label=step.op)
        plates = _reshape_plates(step.plates, np_count=step.np, label=step.op)
        spaixd, spaixi = sp.dskmi2(
            vrtces,
            plates,
            step.finscl,
            int(step.corscl),
            step.worksz,
            step.voxpsz,
            step.voxlsz,
            step.makvtl,
            step.spxisz,
        )

        spaixd_arr = np.asarray(spaixd, dtype=np.float64)
        spaixi_arr = np.asarray(spaixi, dtype=np.int32)
        if step.spaixId is not None:
            state.spatialIndexes[step.spaixId] = (spaixd_arr.copy(), spaixi_arr.copy())

        return StepOutput(op=step.op, value=_summarize_spatial_index(spaixd_arr, spaixi_arr))

    if isinstance(step, StepFileIoDskw02):
        handle_state = _require_handle(
            state,
            step.handleId,
            expected_kind="DAS",
            require_open=True,
        )
        spaixd, spaixi = _require_spatial_index(state, step.spaixId)
        vrtces = _reshape_vrtces(step.vrtces, nv=step.nv, label=step.op)
        plates = _reshape_plates(step.plates, np_count=step.np, label=step.op)
        corpar = np.asarray(step.corpar, dtype=np.float64)

        sp.dskw02(
            handle_state.handle,
            step.center,
            step.surfid,
            step.dclass,
            step.frame,
            step.corsys,
            corpar,
            step.mncor1,
            step.mxcor1,
            step.mncor2,
            step.mxcor2,
            step.mncor3,
            step.mxcor3,
            step.first,
            step.last,
            vrtces,
            plates,
            spaixd,
            spaixi,
        )
        return StepOutput(op=step.op, value=None)

    return None
